src/NuGet.Core/NuGet.Packaging/Rules/TargetFrameworkMissingPlatformVersionRule.cs

```diff
@@ -0,0 +1,172 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Xml;
+using System.Xml.Linq;
+using NuGet.Client;
+using NuGet.Common;
+using NuGet.ContentModel;
+using NuGet.Frameworks;
+using NuGet.Packaging.Core;
+using NuGet.RuntimeModel;
+
+namespace NuGet.Packaging.Rules
+{
+    internal class TargetFrameworkMissingPlatformVersionRule : IPackageRule
+    {
+        // NOTE: We never actually generate messages, only error on failure.
+        public string MessageFormat => "";
+
+        public TargetFrameworkMissingPlatformVersionRule()
+        {
+        }
+
+        public IEnumerable<PackagingLogMessage> Validate(PackageArchiveReader builder)
+        {
+            return Validate(new NuspecReader(builder.GetNuspec()), builder.GetFiles());
+        }
+
+        internal static IEnumerable<PackagingLogMessage> Validate(NuspecReader reader, IEnumerable<string> files)
+        {
+            var bads = (new string[]
+            {
+                ValidateDependencyGroups(reader),
+                ValidateReferenceGroups(reader),
+                ValidateFrameworkAssemblies(reader),
+                ValidateFiles(files)
+            }).Where(err => err != null);
+
+            if (bads.Any())
+            {
+                string items = string.Empty;
+                foreach (string badMessage in bads)
+                {
+                    items += "\n- " + badMessage;
+                }
+                // PackagingLogMessage technically has an Error level, but
+                // that doesn't look like it fails pack, which is what we need
+                // to do in this case.
+                throw new PackagingException(NuGetLogCode.NU1012, string.Format(CultureInfo.CurrentCulture, Strings.MissingTargetPlatformVersions, items));
```

Comment:
- I would create an additional error code, one for error and one for warnings. That way, users can disable the warning at will.
- As far as I know, validation rules (`IPackageRule` based objects) are mean't to analyze package content. If you want to fail pack, I would write that logic in `PackageBuilder` class.